### PR TITLE
EIP-39: monotonic creation height rule

### PR DESCRIPTION
**Pull request by kushti**
In this PR, EIP-39 (https://github.com/ergoplatform/eips/pull/80) is implemented as follows:

* starting from block version #3 (5.0 soft-fork and after), the client is checking that for every output of a transaction, its creation height is not less than max creation height in inputs
* note, only inputs are considered, data inputs are not
* this rule is about soft-fork, as old nodes will continue to validate new blocks. So it will be activated in 5.0 soft-fork (can be merged with existing code even now). Also, the new rule is soft-forkable, so can be switched off (and so, replaced with another rule) via a soft-fork.

### CODE REVIEW

**Review by greenhat — APPROVED** [review]

**Review by MrStahlfelge — APPROVED** [review]

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala [hunk 1]
```diff
@@ -371,6 +374,17 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
     // Cost limit per block
     val maxCost = stateContext.currentParameters.maxBlockCost.toLong
 
+    val blockVersion = stateContext.blockVersion
+
+    // Before 5.0 soft-fork (v3 block version), we are checking only that
+    // creation height in outputs is not non-negative
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala:380]:
```suggestion
    // creation height in outputs is not negative
```

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 2]
```diff
@@ -36,6 +36,52 @@ class ErgoTransactionSpec extends ErgoPropertyTest with ErgoTestConstants {
...
+  }
+
+  private def modifyAsset(boxCandidate: ErgoBoxCandidate,
+                          deltaFn: Long => Long,
+                          idToskip: TokenId): ErgoBoxCandidate = {
+    val assetId = boxCandidate.additionalTokens.find(t => !java.util.Arrays.equals(t._1, idToskip)).get._1
+
+    val tokens = boxCandidate.additionalTokens.map { case (id, amount) =>
+      if (java.util.Arrays.equals(id, assetId)) assetId -> deltaFn(amount) else assetId -> amount
+    }
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala:71]: The search over boxCandidate.additionalTokens happens 3 times. Passing an index in additionalTokens will be both faster and cleaner.

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 3]
```diff
@@ -36,6 +36,52 @@ class ErgoTransactionSpec extends ErgoPropertyTest with ErgoTestConstants {
...
+    new ErgoBoxCandidate(
+      boxCandidate.value,
+      boxCandidate.ergoTree,
+      boxCandidate.creationHeight,
+      tokens,
+      boxCandidate.additionalRegisters)
+  }
+
+  private def checkTx(from: IndexedSeq[ErgoBox], wrongTx: ErgoTransaction): Try[Int] = {
+    wrongTx.statelessValidity().flatMap(_ => wrongTx.statefulValidity(from, emptyDataBoxes, emptyStateContext))
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala:82]: please add comment on why it is called `wrongTx`?

**Review by pragmaxim — APPROVED** [review]

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala (hunk 2 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala:71]: This is testing code

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala (hunk 3 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala:82]: Out of scope of this PR, renamed to tx anyway

**Comment by scalahub** [comment]
> * its creation height is not less than max creation height in inputs

Do you mean "its creation height is not less *or equal to* than max creation height in inputs" ?
If not then we should be able to preserve the same (max) creation height from inputs (cc @greenhat)

**Comment by greenhat** [comment]
> > * its creation height is not less than max creation height in inputs
> 
> Do you mean "its creation height is not less _or equal to_ than max creation height in inputs" ? If not then we should be able to preserve the same (max) creation height from inputs (cc @greenhat)

There are other inputs (with much higher heights) besides the pool box in the update tx.

**Comment by kushti** [comment]
> > * its creation height is not less than max creation height in inputs
> 
> Do you mean "its creation height is not less _or equal to_ than max creation height in inputs" ? If not then we should be able to preserve the same (max) creation height from inputs (cc @greenhat)

not less means equal or greater

**Pull request merged**